chrome/content/zotero/xpcom/recognizePDF.js

```diff
@@ -0,0 +1,463 @@
+/*
+    ***** BEGIN LICENSE BLOCK *****
+    
+    Copyright © 2009 Center for History and New Media
+                     George Mason University, Fairfax, Virginia, USA
+                     http://zotero.org
+    
+    This file is part of Zotero.
+    
+    Zotero is free software: you can redistribute it and/or modify
+    it under the terms of the GNU Affero General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+    
+    Zotero is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU Affero General Public License for more details.
+    
+    You should have received a copy of the GNU Affero General Public License
+    along with Zotero.  If not, see <http://www.gnu.org/licenses/>.
+    
+    ***** END LICENSE BLOCK *****
+*/
+
+Zotero.RecognizePDF = new function () {
+	const OFFLINE_RECHECK_DELAY = 60 * 1000;
+	const RETRY_MAX_INTERVAL = 60 * 1000;
+	
+	this.ROW_QUEUED = 1;
+	this.ROW_PROCESSING = 2;
+	this.ROW_FAILED = 3;
+	this.ROW_SUCCEEDED = 4;
+	
+	this.getRows = getRows;
+	this.getTotal = getTotal;
+	this.getProcessed = getProcessed;
+	this.cancel = cancel;
+	this.canRecognize = canRecognize;
+	this.recognizeItems = recognizeItems;
+	this.addListener = addListener;
+	this.removeListener = removeListener;
+	
+	let _listeners = {};
+	let _rows = [];
+	let _queue = [];
+	let _queueProcessing = false;
+	let _retryNr = 0;
```

Comment:
So if we use `client.makeRequest` maybe for now we don't need any additional retry logic at all. Because, in practice, network connection problems is the main reason for retries anyway.